plugins/events-correlation-engine/src/main/java/org/opensearch/plugin/correlation/events/transport/TransportIndexCorrelationAction.java

```diff
@@ -0,0 +1,498 @@
+    }
+
+    @Override
+    protected void doExecute(Task task, IndexCorrelationRequest request, ActionListener<IndexCorrelationResponse> listener) {
+        AsyncIndexCorrelationAction asyncAction = new AsyncIndexCorrelationAction(request, listener);
+        asyncAction.start();
+    }
+
+    class AsyncIndexCorrelationAction {
+        private final IndexCorrelationRequest request;
+
+        private final ActionListener<IndexCorrelationResponse> listener;
+
+        AsyncIndexCorrelationAction(IndexCorrelationRequest request, ActionListener<IndexCorrelationResponse> listener) {
+            this.request = request;
+            this.listener = listener;
+        }
+
+        void start() {
+            String inputIndex = request.getIndex();
+            String event = request.getEvent();
+
+            NestedQueryBuilder queryBuilder = QueryBuilders.nestedQuery(
+                "correlate",
+                QueryBuilders.matchQuery("correlate.index", inputIndex),
+                ScoreMode.None
+            );
+            SearchSourceBuilder searchSourceBuilder = new SearchSourceBuilder();
+            searchSourceBuilder.query(queryBuilder);
+            searchSourceBuilder.fetchSource(true);
+
+            SearchRequest searchRequest = new SearchRequest();
+            searchRequest.indices(CorrelationRule.CORRELATION_RULE_INDEX);
+            searchRequest.source(searchSourceBuilder);
+
+            client.search(searchRequest, new ActionListener<>() {
+                @Override
+                public void onResponse(SearchResponse response) {
+                    if (response.isTimedOut()) {
+                        onFailures(new OpenSearchStatusException(response.toString(), RestStatus.REQUEST_TIMEOUT));
+                    }
+
+                    Iterator<SearchHit> hits = response.getHits().iterator();
+                    List<CorrelationRule> correlationRules = new ArrayList<>();
+                    while (hits.hasNext()) {
+                        try {
+                            SearchHit hit = hits.next();
+
+                            XContentParser xcp = XContentType.JSON.xContent()
+                                .createParser(xContentRegistry, LoggingDeprecationHandler.INSTANCE, hit.getSourceAsString());
+
+                            CorrelationRule rule = CorrelationRule.parse(xcp);
+                            correlationRules.add(rule);
+                        } catch (IOException e) {
+                            onFailures(e);
+                        }
+                    }
+
+                    prepRulesForCorrelatedEventsGeneration(inputIndex, event, correlationRules);
+                }
+
+                @Override
+                public void onFailure(Exception e) {
+                    onFailures(e);
+                }
+            });
+        }
+
+        private void prepRulesForCorrelatedEventsGeneration(String index, String event, List<CorrelationRule> correlationRules) {
```

Comment:
there are very long method, can break into multiple small private methods

Reply:
can address this refactoring in a follow-up pr.